File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/LogCallback.java
package org.maplibre.nativeffi;

/**
 * Receives process-global MapLibre Native log records.
 *
 * <p>Native code may invoke this callback on logging or worker threads. The callback should return
 * quickly and avoid calling MapLibre APIs. The binding copies each record before invoking Java code
 * and contains callback exceptions so they do not unwind into native code.
 */
@FunctionalInterface
public interface LogCallback {
  /**
   * Handles a copied log record.
   *
   * @return true when the callback consumed the record, false to let native logging handle it
   */
  boolean log(LogRecord record);
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/log/LogCallback.kt
package org.maplibre.nativeffi.log

/**
 * Receives process-global Maplibre Native log records.
 *
 * Native code may invoke this callback on logging or worker threads. The callback should return
 * quickly and avoid calling Maplibre APIs. The binding copies each record before invoking Kotlin
 * code and contains callback exceptions so they do not unwind into native code.
 *
 * Whether a record also reaches MapLibre's platform logger is fixed when the callback is
 * registered, not decided per record, because a host that cannot answer on the producing thread has
 * no way to decide it there.
 */
public fun interface LogCallback {
  /** Receives one log record. */
  public fun log(record: LogRecord)
}
